### DISCUSSION TITLE
[v1] defaultValue being cached

### DISCUSSION THREAD

**Discussion by halljus** [edmundhung/conform#485] [Ideas]
### Describe the bug and the expected behavior

I'm not sure if this is a regression in v1 or if it's now intentional, but in previous versions of Conform leading up to v1, you "fixed" it so `defaultValue` was not cached and thus the form would update when new `defaultValue` was passed in to `useForm` (PR #130, released in Conform 0.6.1). Can we get that behavior back? 🙏 

### Conform version

v1.0.2

### Steps to Reproduce the Bug or Issue

https://stackblitz.com/edit/remix-run-remix-xq61cg?file=app%2Froutes%2F_index.tsx

**Comment by edmundhung** [comment]
This is intentional. But you can update the defaultValue either by `form.reset()` or ~~`form.update()`~~. Why do you need it to be dynamic?

**Comment by halljus** [comment]
> This is intentional. But you can update the defaultValue either by `form.reset()` or `form.update()`. Why do you need it to be dynamic?

We have a scenario where you can choose an entity in a parent route and the form to edit that entity renders in a nested/child route (Outlet). If you change the selected entity, the nested route changes, loader revalidates, the new entity data comes in, and is passed to `defaultValue`. Prior to v1, this just worked. In v1, the form retains the initial values (doesn't update to the new entity values).

We have workarounds, like wrapping the Outlet view (form route) in its own component with a `key` prop or calling `form.reset()` after a navigation, but it feels a bit like a step backwards? Not the end of the world or anything, though.

**Comment by edmundhung** [comment]
Maybe your case is a bit unique with a special route hierarchy such that Remix does re-mount the route. What I usually hear is the exact opposite in which people expect the form to reset when they navigate to the same route with a different `id` (e.g. `/something/a` -> `/something/b`), while Remix preserve the route and so they see the previous form value and have to reset it manually 😅

I don't think this is gonna be helpful as there are many cases it won't work as you expected. There are also implication on how all metadata are derived if we change this behaviour. Maybe you can consider the unique id approach suggested [here](https://conform.guide/api/react/useForm#tips).

**Comment by halljus** [comment]
> Maybe your case is a bit unique with a special route hierarchy such that Remix does re-mount the route. What I usually hear is the exact opposite in which people expect the form to reset when they navigate to the same route with a different `id` (e.g. `/something/a` -> `/something/b`), while Remix preserve the route and so they see the previous form value and have to reset it manually 😅

I might've described it strangely - what you just said is exactly our scenario. 😁 
 
> I don't think this is gonna be helpful as there are many cases it won't work as you expected. There are also implication on how all metadata are derived if we change this behaviour. Maybe you can consider the unique id approach suggested [here](https://conform.guide/api/react/useForm#tips).

Oooh, yeah that might actually work! We'll try it.

**Comment by edmundhung** [comment]
> I might've described it strangely - what you just said is exactly our scenario. 😁

I am just curious, but I thought you are saying you have a case that works before with v0 and  my understanding is that it never works from what I have seen 🤔

**Comment by halljus** [comment]
We tried the data-dependent `id` trick - it did reset the form values, but it messes with focus management (runs after our own useEffect that focuses a particular input once navigation has settled, even though ours comes after useForm in the code) so it won't work for us. Key prop on a wrapper component will, though.

> I am just curious, but I thought you are saying you have a case that works before with v0 and my understanding is that it never works from what I have seen 🤔

Here's the same example using Conform 0.9.1 (it doesn't cache `defaultValue`, works great):

https://stackblitz.com/edit/remix-run-remix-vsjqbc?file=app%2Froutes%2F_index.tsx

**Comment by edmundhung** [comment]
Ah! This only works if you have never change the value on the input. This might be a bug on React because that's the usual expectation with defaultValue: Once the input is mounted, it will never update the value again even default value is changed.

**Comment by halljus** [comment]
Ahh, interesting. So my repro might be flawed and perhaps we're doing something else in our real app that made it work in v0.9. Doesn't matter now, I guess. Good catch, though!

**Comment by edmundhung** [comment]
> We tried the data-dependent id trick - it did reset the form values, but it messes with focus management (runs after our own useEffect that focuses a particular input once navigation has settled, even though ours comes after useForm in the code) so it won't work for us. Key prop on a wrapper component will, though.

The reset happens in a layout effect so it supposes to run earlier than the effect. But I guess you might have delay it because you are waiting for the updated loaded data and the condition you had on the effect makes it run earlier?

I don't have any concern with the `key` approach here. I just wanted to make sure I didn't missed anything.

**Comment by halljus** [comment]
> I don't have any concern with the `key` approach here. I just wanted to make sure I didn't missed anything.

We'll stick with the `key` approach then, as it's more clear to everyone anyway.
